app/services/magic_link_sign_up_flow.rb

```diff
@@ -0,0 +1,25 @@
+class MagicLinkSignUpFlow
+  def self.begin(candidate)
+    magic_link_token = MagicLinkToken.new
+    candidate.update!(magic_link_token: magic_link_token.encrypted)
+    AuthenticationMailer.sign_up_email(to: candidate.email_address, token: magic_link_token.raw).deliver!
+    candidate.update!(magic_link_token_sent_at: Time.now)
+  end
+
+  def self.find_authenticated_candidate(raw_token:)
+    token = MagicLinkToken.from_raw(raw_token)
+    Candidate.find_by(magic_link_token: token)
+  end
+
+  class MagicLinkToken
+    attr_reader :encrypted, :raw
+
+    def initialize
+      @raw, @encrypted = Devise.token_generator.generate(Candidate, :magic_link_token)
+    end
+
+    def self.from_raw(token)
+      Devise.token_generator.digest(Candidate, :magic_link_token, token)
+    end
+  end
+end
```

Comment:
Even if this is into the `/services` folder, this is not a service (see https://medium.com/selleo/essential-rubyonrails-patterns-part-1-service-objects-1af9f9573ca1)

- It is fine that is it not a service, it looks a bit complex to be a plain service.
- Instead of forcing it to be a Service, I'd move it to a different folder, maybe `/lib` or `/entities`.